File: expense_tracker.py
import csv
from datetime import datetime

FILE = "expenses.csv"

# Add Expense
def add_expense():
    amount = input("Enter amount: ")
    category = input("Enter category (Food/Travel/Shopping/etc): ")
    date = datetime.now().strftime("%Y-%m-%d")

    with open(FILE, "a", newline="") as file:
        writer = csv.writer(file)
        writer.writerow([amount, category, date])

    print(" Expense added successfully!")

# View Expenses
def view_expenses():
    try:
        with open(FILE, "r") as file:
            reader = csv.reader(file)
            print("\n--- All Expenses ---")
            for row in reader:
                print(f"Amount: {row[0]}, Category: {row[1]}, Date: {row[2]}")
    except FileNotFoundError:
        print("No expenses found!")

# Total Expense
def total_expense():
    total = 0
    try:
        with open(FILE, "r") as file:
            reader = csv.reader(file)
            for row in reader:
                total += float(row[0])
        print(f"Total Expense: {total}")
    except FileNotFoundError:
        print("No data found!")

# Category wise Total
def category_wise():
    data = {}
    try:
        with open(FILE, "r") as file:
            reader = csv.reader(file)
            for row in reader:
                category = row[1]
                amount = float(row[0])

                if category in data:
                    data[category] += amount
                else:
                    data[category] = amount

        print("\n--- Category-wise Expense ---")
        for cat, amt in data.items():
            print(f"{cat}: {amt}")

    except FileNotFoundError:
        print("No data found!")

# Delete Expense
def delete_expense():
    try:
        with open(FILE, "r") as file:
            reader = list(csv.reader(file))
            if not reader:
                print("No expenses to delete")
                return
            print("\n---Expense List---")
            for i, row in enumerate(reader):
                print(f"{i}. Amount: {row[0]}, Category: {row[1]}, Date: {row[2]}")
                index = int(input("Enter index to delete :"))
                if index < 0 or index >= len(reader):
                    print("Invalid Index")
                    return
                reader.pop(index)
                with open(FILE, "w", newline="") as file:
                    writer = csv.writer(file)
                    writer.writerows(reader)
                print("Expense deleted successfully")
    except FileNotFoundError:
        print("No data found")
    except ValueError:
        print("Invalid input, Enter a number")


def monthly_expense():
    from datetime import datetime
    current_month = datetime.now().strftime("%Y-%m")

    total = 0

    try:
        with open(FILE, "r") as file:
            reader = csv.reader(file)

            print("\n--- Current Month Expenses ---")
            for row in reader:
                if row[2].startswith(current_month):
                    print(f"Amount: {row[0]}, Category: {row[1]}, Date: {row[2]}")
                    total += float(row[0])

        print(f"\nTotal this month: {total}")

    except FileNotFoundError:
        print("No data found!")


def show_graph():
    import matplotlib.pyplot as plt

    data = {}

    try:
        with open(FILE, "r") as file:
            reader = csv.reader(file)

            for row in reader:
                category = row[1]
                amount = float(row[0])

                if category in data:
                    data[category] += amount
                else:
                    data[category] = amount

        categories = list(data.keys())
        amounts = list(data.values())

        plt.bar(categories, amounts)
        plt.xlabel("Category")
        plt.ylabel("Amount")
        plt.title("Expense Analysis")
        plt.show()

    except FileNotFoundError:
        print("No data found!")

# Menu
def menu():
    while True:
        print("\n===== Expense Tracker =====")
        print("1. Add Expense")
        print("2. View Expenses")
        print("3. Total Expense")
        print("4. Category-wise Expense")
        print("5. Delete Expense")
        print("6. Monthly Expense")
        print("7. Show Graph")
        print("8. Exit")

        choice = input("Enter your choice: ")

        if choice == "1":
            add_expense()
        elif choice == "2":
            view_expenses()
        elif choice == "3":
            total_expense()
        elif choice == "4":
            category_wise()
        elif choice == "5":
            delete_expense()
        elif choice == "6":
            monthly_expense()
        elif choice == "7":
            show_graph()
        elif choice == "8":
            print("👋 Exiting...")
            break
        else:
            print("Invalid choice, try again!")

# Run program
menu()

File: expenseTracker.java
import java.io.*;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.nio.file.*;

public class expenseTracker {

    static final String FILE = "expenses.csv";
    static Scanner sc = new Scanner(System.in);

    // Add Expense
    static void addExpense() {
        try {
            System.out.print("Enter amount: ");
            String amount = sc.nextLine();

            System.out.print("Enter category (Food/Travel/Shopping/etc): ");
            String category = sc.nextLine();

            String date = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM-dd"));

            FileWriter fw = new FileWriter(FILE, true);
            fw.write(amount + "," + category + "," + date + "\n");
            fw.close();

            System.out.println("Expense added successfully!");

        } catch (IOException e) {
            System.out.println("Error writing file");
        }
    }

    // View Expenses
    static void viewExpenses() {
        try {
            File file = new File(FILE);

            if (!file.exists()) {
                System.out.println("No expenses found!");
                return;
            }

            BufferedReader br = new BufferedReader(new FileReader(FILE));

            String line;

            System.out.println("\n--- All Expenses ---");

            while ((line = br.readLine()) != null) {
                String[] row = line.split(",");

                System.out.println("Amount: " + row[0]
                        + ", Category: " + row[1]
                        + ", Date: " + row[2]);
            }

            br.close();

        } catch (IOException e) {
            System.out.println("Error reading file");
        }
    }

    // Total Expense
    static void totalExpense() {
        double total = 0;

        try {
            File file = new File(FILE);

            if (!file.exists()) {
                System.out.println("No data found!");
                return;
            }

            BufferedReader br = new BufferedReader(new FileReader(FILE));

            String line;

            while ((line = br.readLine()) != null) {
                String[] row = line.split(",");
                total += Double.parseDouble(row[0]);
            }

            br.close();

            System.out.println("Total Expense: " + total);

        } catch (IOException e) {
            System.out.println("Error reading file");
        }
    }

    // Category Wise Expense
    static void categoryWise() {
        HashMap<String, Double> data = new HashMap<>();

        try {
            File file = new File(FILE);

            if (!file.exists()) {
                System.out.println("No data found!");
                return;
            }

            BufferedReader br = new BufferedReader(new FileReader(FILE));

            String line;

            while ((line = br.readLine()) != null) {
                String[] row = line.split(",");

                String category = row[1];
                double amount = Double.parseDouble(row[0]);

                data.put(category, data.getOrDefault(category, 0.0) + amount);
            }

            br.close();

            System.out.println("\n--- Category-wise Expense ---");

            for (String cat : data.keySet()) {
                System.out.println(cat + ": " + data.get(cat));
            }

        } catch (IOException e) {
            System.out.println("Error reading file");
        }
    }

    // Delete Expense
    static void deleteExpense() {
        try {
            File file = new File(FILE);

            if (!file.exists()) {
                System.out.println("No data found!");
                return;
            }

            List<String> lines = Files.readAllLines(Paths.get(FILE));

            if (lines.isEmpty()) {
                System.out.println("No expenses to delete");
                return;
            }

            System.out.println("\n--- Expense List ---");

            for (int i = 0; i < lines.size(); i++) {
                String[] row = lines.get(i).split(",");

                System.out.println(i + ". Amount: " + row[0]
                        + ", Category: " + row[1]
                        + ", Date: " + row[2]);
            }

            System.out.print("Enter index to delete: ");
            int index = Integer.parseInt(sc.nextLine());

            if (index < 0 || index >= lines.size()) {
                System.out.println("Invalid Index");
                return;
            }

            lines.remove(index);

            BufferedWriter bw = new BufferedWriter(new FileWriter(FILE));

            for (String line : lines) {
                bw.write(line);
                bw.newLine();
            }

            bw.close();

            System.out.println("Expense deleted successfully");

        } catch (IOException e) {
            System.out.println("Error reading file");
        } catch (NumberFormatException e) {
            System.out.println("Invalid input, Enter a number");
        }
    }

    // Monthly Expense
    static void monthlyExpense() {
        String currentMonth = LocalDate.now().format(DateTimeFormatter.ofPattern("yyyy-MM"));

        double total = 0;

        try {
            File file = new File(FILE);

            if (!file.exists()) {
                System.out.println("No data found!");
                return;
            }

            BufferedReader br = new BufferedReader(new FileReader(FILE));

            String line;

            System.out.println("\n--- Current Month Expenses ---");

            while ((line = br.readLine()) != null) {
                String[] row = line.split(",");

                if (row[2].startsWith(currentMonth)) {
                    System.out.println("Amount: " + row[0]
                            + ", Category: " + row[1]
                            + ", Date: " + row[2]);

                    total += Double.parseDouble(row[0]);
                }
            }

            br.close();

            System.out.println("\nTotal this month: " + total);

        } catch (IOException e) {
            System.out.println("Error reading file");
        }
    }

    // Menu
    static void menu() {
        while (true) {

            System.out.println("\n===== Expense Tracker =====");
            System.out.println("1. Add Expense");
            System.out.println("2. View Expenses");
            System.out.println("3. Total Expense");
            System.out.println("4. Category-wise Expense");
            System.out.println("5. Delete Expense");
            System.out.println("6. Monthly Expense");
            System.out.println("7. Exit");

            System.out.print("Enter your choice: ");
            String choice = sc.nextLine();

            switch (choice) {

                case "1":
                    addExpense();
                    break;

                case "2":
                    viewExpenses();
                    break;

                case "3":
                    totalExpense();
                    break;

                case "4":
                    categoryWise();
                    break;

                case "5":
                    deleteExpense();
                    break;

                case "6":
                    monthlyExpense();
                    break;

                case "7":
                    System.out.println("Exiting...");
                    return;

                default:
                    System.out.println("Invalid choice, try again!");
            }
        }
    }

    // Main Method
    public static void main(String[] args) {
        menu();
    }
}